lib/ash/actions/aggregate.ex

```diff
@@ -6,101 +6,95 @@ defmodule Ash.Actions.Aggregate do
-                       }),
-                     {:ok, result} <-
-                       Ash.DataLayer.run_aggregate_query(
-                         data_layer_query,
-                         aggregates,
-                         query.resource
-                       ) do
-                  {:cont, {:ok, Map.merge(acc, result)}}
-                else
-                  {:error, error} ->
-                    {:halt, {:error, error}}
-                end
-              end
+
+    aggregates
+    |> Enum.group_by(fn
+      %Ash.Query.Aggregate{} = aggregate ->
+        agg_authorize? = aggregate.authorize? && opts[:authorize?]
+
+        read_action =
+          aggregate.read_action || (query.action && query.action.name) ||
+            Ash.Resource.Info.primary_action!(query.resource, :read).name
+
+        {agg_authorize?, read_action}
+
+      {_name, _kind} ->
+        {!!opts[:authorize?],
+         opts[:read_action] || opts[:action] || (query.action && query.action.name) ||
+           Ash.Resource.Info.primary_action!(query.resource, :read).name}
+
+      {_name, _kind, agg_opts} ->
+        authorize? =
+          Keyword.get(agg_opts, :authorize?, true) && opts[:authorize?]
+
+        {authorize?,
+         agg_opts[:read_action] || opts[:read_action] || agg_opts[:action] || opts[:action] ||
+           (query.action && query.action.name) ||
+           Ash.Resource.Info.primary_action!(query.resource, :read).name}
+    end)
+    |> Enum.reduce_while({:ok, %{}}, fn
+      {{agg_authorize?, read_action}, aggregates}, {:ok, acc} ->
+        query =
+          if query.__validated_for_action__ == read_action do
+            query
+          else
+            Ash.Query.for_read(query, read_action, %{},
+              tenant: opts[:tenant],
+              actor: opts[:actor],
+              authorize?: opts[:authorize?]
+            )
+          end
+
+        query = %{query | domain: domain}
+
+        Ash.Tracer.span :action,
+                        Ash.Domain.Info.span_name(query.domain, query.resource, :aggregate),
+                        opts[:tracer] do
+          metadata = %{
+            domain: query.domain,
+            resource: query.resource,
+            resource_short_name: Ash.Resource.Info.short_name(query.resource),
+            aggregates: List.wrap(aggregates),
+            actor: opts[:actor],
+            tenant: opts[:tenant],
+            action: read_action,
+            authorize?: opts[:authorize?]
+          }
+
+          Ash.Tracer.telemetry_span [
+                                      :ash,
+                                      Ash.Domain.Info.short_name(query.domain),
+                                      :aggregate
+                                    ],
+                                    metadata do
+            Ash.Tracer.set_metadata(opts[:tracer], :action, metadata)
+
+            with {:ok, query} <- Ash.Actions.Read.handle_multitenancy(query),
```

Comment:
This is a messy diff but basically I removed

```elixir
query = Ash.Actions.Read.handle_attribute_multitenancy(query)

if query.valid do
  with :ok <- Ash.Actions.Read.validate_multitenancy(query) do
```

which removed a couple of layers of indentation, and then I've put `handle_multitenancy` here since it now depends from the specific `read_action` used for each aggregate